qiskit/primitives/containers/bindings_array.py

```diff
@@ -33,67 +33,51 @@
-    ``[vals0, vals1, ...]`` in the bindings it specifies. This allows flexibility about whether
-    values for different parameters are stored in one big array, or across several smaller
-    arrays. It also allows different parameters to use different dtypes.
-
-    .. code-block:: python
-
-        # 0-d array (i.e. only one binding)
-        BindingsArray([1, 2, 3], {"a": 4, ("b", "c"): [5, 6]})
-
-        # single array, last index is parameters
-        BindingsArray(np.empty((10, 10, 100)))
-
-        # multiple arrays, where each last index is parameters. notice that it's smart enough to
-        # figure out that a missing last dimension corresponds to a single parameter.
-        BindingsArray(
-            [np.empty((10, 10, 100)), np.empty((10, 10)), np.empty((10, 10, 20), dtype=complex)],
-            {("c", "a"): np.empty((10, 10, 2)), "b": np.empty((10, 10))}
-        )
+        A single parameter binding set provides numeric values to bind to a circuit with free
+        :class:`qiskit.circuit.Parameter`\s. An instance of this class stores an array-valued collection
+        of such sets. The simplest example is a 0-d array consisting of a single parameter binding set,
+        whereas an n-d array of parameter binding sets represents an n-d sweep over values.
+
+        The storage format is a dictionary of arrays attached to parameters, ``{params_0: values_0,
+        ...}``. A convention is used where the last dimension of each array indexes (a subset of)
+        circuit parameters. For example, if the last dimension of ``values_0`` is 25, then it represents
+        an array of possible binding values for the 25 distinct parameters ``params_0``, where its
+        leading shape is the array :attr:`~.shape` of its binding array. This allows flexibility about
+        whether values for different parameters are stored in one big array, or across several smaller
+        arrays. It also allows different parameters to use different dtypes.
+
+        .. code-block:: python
+
+            # 0-d array (i.e. only one binding)
+            BindingsArray({"a": 4, ("b", "c"): [5, 6]})
+    ß
+            # single array, last index is parameters
+            parameters = tuple(f"a{idx}" for idx in range(100))
+            BindingsArray({parameters: np.empty((10, 10, 100))})
+
+            # multiple arrays, where each last index is parameters. notice that it's smart enough to
+            # figure out that a missing last dimension corresponds to a single parameter.
+            BindingsArray(
+                {("c", "a"): np.empty((10, 10, 2)), "b": np.empty((10, 10), dtype=int)}
+            )
     """
-    __slots__ = ("_vals", "_kwvals")
 
     def __init__(
         self,
-        vals: ArrayLike | Iterable[ArrayLike] | None = None,
-        kwvals: Mapping[ParameterLike, Iterable[ParameterValueType]] | ArrayLike | None = None,
+        data: Mapping[ParameterLike, Iterable[ParameterValueType]] | ArrayLike | None = None,
```

Comment:
Is `ArrayLike` allowed as `data`? The following case raises an error.

```python
In [5]: ba = BindingsArray([])
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[5], line 1
----> 1 ba = BindingsArray([])

File ~/tasks/4_2024/qiskit/terra/qiskit/primitives/containers/bindings_array.py:98, in BindingsArray.__init__(self, data, shape)
     92     self._data = {}
     93 else:
     94     self._data = {
     95         _format_key((p,))
     96         if isinstance(p, Parameter)
     97         else _format_key(p): np.array(val, copy=False)
---> 98         for p, val in data.items()
     99     }
    101 self._shape = _infer_shape(self._data) if shape is None else shape_tuple(shape)
    102 self._num_parameters = None

AttributeError: 'list' object has no attribute 'items'
```

Reply:
@t-imamichi No, with this change its not. A zero param binding array would need to be initialized as `BindingsArray()`, `BindingsArray(None)` or `BindingsArray({})`